spidermon/contrib/monitors/mixins/validation.py

```diff
@@ -84,125 +84,259 @@ def __init__(self, stats, prefix=None):
+                self.validation.fields[field_name].errors["missing_required_field"].percent
+            )
+            if missing_percent > allowed_percent:
+                msg = self._get_msg_for_missing_required_percent(field_name, missing_percent, allowed_percent)
+                msgs.append(msg)
+        if msgs:
+            msgs.insert(0, 'Required fields are missing:')
+            self.fail('\n'.join(msgs))
 
     def check_missing_required_field_percent(self, field_name, allowed_percent=0):
+        """
+        Checks that the number of "missing_required_field" errors for the ``field_name`` field divided by the number
+        of items is less or equal than ``allowed_percent``.
+        """
         missing_percent = (
             self.validation.fields[field_name].errors["missing_required_field"].percent
         )
-        self._check_missing_required_percent(missing_percent, allowed_percent)
-
-    def _check_missing_required_percent(self, missing_percent, allowed_percent=0):
-        self.assertLessEqual(
-            missing_percent,
-            allowed_percent,
-            msg="{percent}% of required fields are missing!{threshold_info}".format(
-                percent=missing_percent * 100,
-                threshold_info=(" (maximum allowed %.0f%%)" % (allowed_percent * 100))
-                if allowed_percent > 0
-                else "",
-            ),
+        msg = self._get_msg_for_missing_required_percent(field_name, missing_percent, allowed_percent)
+        self.assertLessEqual(missing_percent, allowed_percent, msg)
+
+    @staticmethod
+    def _get_msg_for_missing_required_percent(field_name, missing_percent, allowed_percent):
+        msg = "{percent}% of required field {field} are missing!{threshold_info}".format(
+            percent=missing_percent * 100,
+            field=field_name,
+            threshold_info=(" (maximum allowed %.0f%%)" % (allowed_percent * 100))
+            if allowed_percent > 0
+            else "",
         )
+        return msg
 
     def check_fields_errors(self, field_names=None, errors=None, allowed_count=0):
-        if not field_names:
+        """
+        Checks that the number of errors for the ``field_names`` fields is less or equal than ``allowed_count`` and
+        raises an error with all problematic fields.
+        If ``field_names`` is None and ``new_behavior`` is True, checks all fields.
+        If ``field_names`` is None and ``new_behavior`` is False, checks that the total number of errors is less or
+        equal than ``allowed_count``.
+        """
+        if not self.new_behavior and not field_names:
+            # TODO: deprecate this
             errors_count = self.validation.errors.count
-            self._check_field_errors(errors_count, allowed_count)
-        else:
-            for field_name in field_names:
-                self.check_field_errors(field_name, errors, allowed_count)
+            self.assertLessEqual(
```

Comment:
All the methods look quite similar, this first `if` statement then a `for` loop.
Maybe, if you change this assertion to an `if` statement followed by `self.fail` (as below) it would give some insight on how to extract this behavior in a method.
Then, these functions could be only a first check and preparing the proper inputs checks for this method.
More of an idea for refactoring that can be done later :smile: